<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>ExOutBench Demonstration • ExOutBench</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="ExOutBench Demonstration">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ExOutBench</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/exoutbench_tutorial.html">ExOutBench Demonstration</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>ExOutBench Demonstration</h1>
                        <h4 class="author">Jonah Einson</h4>
            
      
      
      <div class="hidden name"><code>exoutbench_tutorial.Rmd</code></div>

    </div>

    
    
<pre><code>## ── Attaching packages ─────────────────────────────────── tidyverse 1.2.1 ──</code></pre>
<pre><code>## ✔ ggplot2 3.1.1       ✔ purrr   0.3.2  
## ✔ tibble  2.1.1       ✔ dplyr   0.8.0.1
## ✔ tidyr   0.8.3       ✔ stringr 1.4.0  
## ✔ readr   1.3.1       ✔ forcats 0.4.0</code></pre>
<pre><code>## ── Conflicts ────────────────────────────────────── tidyverse_conflicts() ──
## ✖ dplyr::filter() masks stats::filter()
## ✖ dplyr::lag()    masks stats::lag()</code></pre>
<p><code>ExOutBench</code> is a tool used to benchmark the performance expression outlier calling methods in prioritizing genes whose expression is associated with rare genetic variation. For each function in the package the input is:</p>
<ul>
<li>A file with rare variant information. Which individual-gene pairs have a rare variant? This is provided, or pre-compiled beforehand.</li>
<li>A file with the expresion outlier status of the same individual-gene pair. This is the thing you’re trying to test. The file with variant information (called <code>outlier.calls</code>) is formatted like this:</li>
</ul>
<table class="table">
<thead><tr class="header">
<th>SampleName</th>
<th>GeneID</th>
<th>chr</th>
<th>start</th>
<th>end</th>
<th>consdetail</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>IndvA</td>
<td>ENSG00000186092</td>
<td>chr1</td>
<td>64904</td>
<td>64905</td>
<td>upstream</td>
</tr>
<tr class="even">
<td>IndvA</td>
<td>ENSG00000230021</td>
<td>chr1</td>
<td>631495</td>
<td>631496</td>
<td>upstream</td>
</tr>
<tr class="odd">
<td>IndvA</td>
<td>ENSG00000230021</td>
<td>chr1</td>
<td>666203</td>
<td>666204</td>
<td>intron,non_coding</td>
</tr>
<tr class="even">
<td>IndvA</td>
<td>ENSG00000230021</td>
<td>chr1</td>
<td>666398</td>
<td>666399</td>
<td>intron,non_coding</td>
</tr>
<tr class="odd">
<td>IndvB</td>
<td>ENSG00000225972</td>
<td>chr1</td>
<td>631495</td>
<td>631496</td>
<td>downstream</td>
</tr>
<tr class="even">
<td>IndvB</td>
<td>ENSG00000225630</td>
<td>chr1</td>
<td>631495</td>
<td>631496</td>
<td>downstream</td>
</tr>
<tr class="odd">
<td>IndvB</td>
<td>ENSG00000237973</td>
<td>chr1</td>
<td>631495</td>
<td>631496</td>
<td>non_coding_exon</td>
</tr>
<tr class="even">
<td>IndvB</td>
<td>ENSG00000229344</td>
<td>chr1</td>
<td>631495</td>
<td>631496</td>
<td>upstream</td>
</tr>
</tbody>
</table>
<p>Every line is an individual-gene pair and a rare variant associated with that gene. There can be multiple rare variants associated with one individual-gene pair, but if an individual-gene pair does not carry any rare variants within the specified window, it will be excluded from the table.</p>
<p>I have compiled this file for European individuals in the GTEx v8 cohort. It lists all rare variants (MAF &lt; 1% in GnomAD and in GTEx) within 10 kb upstream of the gene and within the gene body. The file will have to be shared internally, since it contains personal genotype data.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">rare.variants &lt;-<span class="st"> </span><span class="kw">read_tsv</span>(
  <span class="st">"/gpfs/commons/groups/lappalainen_lab/jeinson/data/ExOutBench_data/all_rare_variants_SNPs_10kb_genebody_w_consdetail_no_NA.tsv"</span>, 
  <span class="dt">progress =</span> F)</code></pre></div>
<pre><code>## Parsed with column specification:
## cols(
##   SampleName = col_character(),
##   GeneID = col_character(),
##   chr = col_character(),
##   start = col_double(),
##   end = col_double(),
##   consdetail = col_character()
## )</code></pre>
<p>The test also takes in expression outlier calls from any method you might want to benchmark. The formatted is as such, where every row is an individual-gene pair with a corresponding “outlier score”:</p>
<table class="table">
<thead><tr class="header">
<th align="left">GeneID</th>
<th align="left">SampleName</th>
<th align="right">outlier.score</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">ENSG00000000971</td>
<td align="left">GTEX-1117F</td>
<td align="right">0.9098267</td>
</tr>
<tr class="even">
<td align="left">ENSG00000001561</td>
<td align="left">GTEX-1117F</td>
<td align="right">0.9605523</td>
</tr>
<tr class="odd">
<td align="left">ENSG00000001617</td>
<td align="left">GTEX-1117F</td>
<td align="right">0.9692674</td>
</tr>
<tr class="even">
<td align="left">ENSG00000002549</td>
<td align="left">GTEX-1117F</td>
<td align="right">0.9269707</td>
</tr>
<tr class="odd">
<td align="left">ENSG00000002822</td>
<td align="left">GTEX-1117F</td>
<td align="right">0.8792655</td>
</tr>
<tr class="even">
<td align="left">ENSG00000002834</td>
<td align="left">GTEX-1117F</td>
<td align="right">0.9964071</td>
</tr>
</tbody>
</table>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">tiss.outlier.scores &lt;-
<span class="st">  </span><span class="kw">read_csv</span>(<span class="st">"/gpfs/commons/groups/lappalainen_lab/jeinson/projects/aneva-dot/analysis_v8/ANEVA-DOT-pipeline/Filtering/combined.ad.scores.in.MSCLSK.tsv"</span>) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">rename</span>(<span class="st">"GeneID"</span> =<span class="st"> "X1"</span>) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">gather</span>(SampleName, DOT.score, <span class="op">-</span>GeneID) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/filter">filter</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/complete.cases">complete.cases</a></span>(.)) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">rename</span>(<span class="st">"outlier.score"</span> =<span class="st"> "DOT.score"</span>)</code></pre></div>
<pre><code>## Warning: Missing column names filled in: 'X1' [1]</code></pre>
<pre><code>## Parsed with column specification:
## cols(
##   .default = col_double(),
##   X1 = col_character()
## )</code></pre>
<pre><code>## See spec(...) for full column specifications.</code></pre>
<p>!! The column names in the outlier call file and the rare variant call file must match! The package joins them together then performs enrichment analyses.</p>
<p>Right now, this package is able to calculate enrichment by</p>
<ul>
<li>
<strong>Outlier score threshold:</strong> This indicates how likely it is to observe a rare variant given outlier status, increasingly stringent outlier score thresholds.</li>
<li>
<strong>Variant annotation:</strong> This shows the relative risk of a certain type of rare variant causing an expression outlier to occur, under a fixed threshold.</li>
</ul>
<p>In the future, I will add support for showing enrichment of rare variants based on their distance to outlier genes, and enrichment of rare variants given their population frequency.</p>
<p>By default, each function removes all individual-gene pairs from genes that never occur as outliers. This functionality can be turned off by setting <code>limit.to.genes.w.outliers</code> to <code>FALSE</code>. The functions produce a data frame, and plots the results automatically by default. See the package documentation for more options and descriptions.</p>
<p>Run the enrichment by significance pipeline!</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">enrichment.by.significance.output &lt;-<span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="../reference/enrichment_by_significance.html">enrichment_by_significance</a></span>(<span class="dt">outlier.calls =</span> tiss.outlier.scores, <span class="dt">rare.variants =</span> rare.variants, <span class="dt">draw.plot =</span> T)</code></pre></div>
<pre><code>## Only considering genes with at least one outlier at 0.05 
## Checking to only include individuals that have genotype data!!</code></pre>
<pre><code>## Joining, by = c("GeneID", "SampleName")</code></pre>
<pre><code>## Calculating enrichment scores</code></pre>
<p><img src="exoutbench_tutorial_files/figure-html/unnamed-chunk-4-1.png" width="576"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">enrichment.by.significance.output</code></pre></div>
<pre><code>##      ratio  lower.q  upper.q n.outliers.w.var n.outliers
## 1 1.214516 1.184871 1.244902             1880       2673
## 2 1.292598 1.253777 1.332621             1043       1393
## 3 1.353310 1.307331 1.400906              697        889
## 4 1.400594 1.348621 1.454569              508        626
## 5 1.434031 1.377525 1.492854              403        485
## 6 1.478486 1.416925 1.542723              305        356
##   n.non.outliers.w.var n.non.outliers   sig
## 1               364503         629427 5e-02
## 2               365340         630707 1e-02
## 3               365686         631211 1e-03
## 4               365875         631474 1e-04
## 5               365980         631615 1e-05
## 6               366078         631744 1e-07</code></pre>
<p>Run the enrichment by annotation pipeline!</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">enrichment.by.annotation.out &lt;-<span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="../reference/enrichment_by_annotation.html">enrichment_by_annotation</a></span>(<span class="dt">outlier.calls =</span> tiss.outlier.scores, <span class="dt">rare.variants =</span> rare.variants, <span class="dt">draw.plot =</span> T)</code></pre></div>
<pre><code>## Warning in if (!(c("SampleName", "GeneID", "chr", "start", "consdetail")
## %in% : the condition has length &gt; 1 and only the first element will be used</code></pre>
<pre><code>## Only considering genes with at least one outlier at 0.05 
## Checking to only include individuals that have genotype data!!
## Combining outlier calls with rare variant information</code></pre>
<pre><code>## Joining, by = c("GeneID", "SampleName")</code></pre>
<pre><code>## Calculating enrichment scores</code></pre>
<p><img src="exoutbench_tutorial_files/figure-html/unnamed-chunk-6-1.png" width="576"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">enrichment.by.annotation.out</code></pre></div>
<pre><code>##         ratio     lower.q    upper.q n.outliers.w.var n.outliers
## 1    1.075880   1.0350329   1.118339             1312       2673
## 2    1.608611   0.9335467   2.771826               13       2673
## 3    1.973116   1.7418131   2.235136              228       2673
## 4    9.078017   7.5648357  10.893878              115       2673
## 5    1.299335   1.1782949   1.432810              351       2673
## 6    1.050469   0.9430467   1.170127              295       2673
## 7    3.424950   2.7594900   4.250887               81       2673
## 8    1.181876   0.8868263   1.575089               46       2673
## 9    2.250708   1.9618846   2.582050              191       2673
## 10   1.967354   1.6505084   2.345024              120       2673
## 11   2.994491   1.6541396   5.420930               11       2673
## 12  29.969656  13.6612328  65.746650                7       2673
## 13 140.227203 110.5887923 177.808873              106       2673
## 14  97.636336  65.5794661 145.363400               34       2673
## 15  27.703044   6.4035126 119.849630                2       2673
## 16  21.406897   2.7646457 165.755509                1       2673
## 17   0.000000   0.0000000        NaN                0       2673
##    n.non.outliers.w.var n.non.outliers            anno
## 1                287155         629427          intron
## 2                  1903         629427             NMD
## 3                 27210         629427     3_prime_UTR
## 4                  2983         629427          splice
## 5                 63611         629427        upstream
## 6                 66128         629427      downstream
## 7                  5569         629427     5_prime_UTR
## 8                  9165         629427      non_coding
## 9                 19983         629427        missense
## 10                14363         629427      synonymous
## 11                  865         629427 non_coding_exon
## 12                   55         629427 splice_acceptor
## 13                  178         629427     stop_gained
## 14                   82         629427    splice_donor
## 15                   17         629427      start_lost
## 16                   11         629427       stop_lost
## 17                    6         629427   stop_retained</code></pre>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by .</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
